lib/Transforms/Tapir/PTXABI.cpp

```diff
@@ -0,0 +1,725 @@
+#include "llvm/Target/TargetMachine.h"           
+#include "llvm/Support/TargetRegistry.h"    
+#include "llvm/IR/LegacyPassManager.h"
+
+#include <iostream>
+#include <set>
+#include <sstream>
+
+#define np(X)                                                            \
+ std::cout << __FILE__ << ":" << __LINE__ << ": " << __PRETTY_FUNCTION__ \
+           << ": " << #X << " = " << (X) << std::endl
+
+#include <iostream>
+#include <set>
+#include <sstream>
+
+using namespace llvm;
+
+namespace{
+
+  template<class F>
+  Function* getFunction(Module& M, const char* name){
+    return cast<Function>(M.getOrInsertFunction(name,
+      TypeBuilder<F, false>::get(M.getContext())));
+  } 
+
+  template<class B>
+  Value* convertInteger(B& b, Value* from, Value* to, const std::string& name){
+    auto ft = dyn_cast<IntegerType>(from->getType());
+    assert(ft && "expected from type as integer type");
+
+    auto tt = dyn_cast<IntegerType>(to->getType());
+    assert(tt && "expected to type as integer type");
+
+    if(ft->getBitWidth() > tt->getBitWidth()){
+      return b.CreateTrunc(from, tt, name);
+    }
+    else if(ft->getBitWidth() < tt->getBitWidth()){
+      return b.CreateZExt(from, tt, name);
+    }
+
+    return from;
+  }
+  
+} // namespace
+
+
+//##############################################################################
+
+PTXABI::PTXABI() {}
+
+/// \brief Get/Create the worker count for the spawning function.
+Value *PTXABI::GetOrCreateWorker8(Function &F) {
+  Module *M = F.getParent(); 
+  LLVMContext& C = M->getContext();
+  return ConstantInt::get(C, APInt(16, 8));
+}
+
+void PTXABI::createSync(SyncInst &SI, ValueToValueMapTy &DetachCtxToStackFrame) {
+}
+
+Function *PTXABI::createDetach(DetachInst &detach,
+                               ValueToValueMapTy &DetachCtxToStackFrame,
+                               DominatorTree &DT, AssumptionCache &AC) {
+  BasicBlock *detB = detach.getParent();
+  // unused -- Function &F = *(detB->getParent());
+
+  BasicBlock *Spawned  = detach.getDetached();
+  BasicBlock *Continue = detach.getContinue();
+
+  // unused -- Module *M = F.getParent();
+
+  CallInst *cal = nullptr;
+  Function *extracted = extractDetachBodyToFunction(detach, DT, AC, &cal);
+  //extracted = formatFunctionToTask(extracted, cal);
+
+  // Replace the detach with a branch to the continuation.
+  BranchInst *ContinueBr = BranchInst::Create(Continue);
+  ReplaceInstWithInst(&detach, ContinueBr);
+
+  // Rewrite phis in the detached block.
+  {
+    BasicBlock::iterator BI = Spawned->begin();
+    while (PHINode *P = dyn_cast<PHINode>(BI)) {
+      P->removeIncomingValue(detB);
+      ++BI;
+    }
+  }
+  return extracted;
+}
+
+void PTXABI::preProcessFunction(Function &F) {
+}
+
+void PTXABI::postProcessFunction(Function &F) {
+}
+
+void PTXABI::postProcessHelper(Function &F) {
+}
+
+bool PTXABI::processMain(Function &F) {
+  return true;
+}
+
+bool PTXABILoopSpawning::processLoop(){
```

Comment:
A bunch of the code in this method seems to recreate the functionality of the methods in Transforms/Tapir/Outline.cpp, which were built specifically to handle function outlining of this nature and address lots of hairy corner cases.  I would prefer that those functions be used to reduce code and functionality duplication.